### PR TITLE
626 parallel distribution

### PR DESCRIPTION
**Pull request by oerc0122**
Add capability for parallel framework to automatically distribute standard MATLAB arrays under 1D decomposition. More complex decompositions can be performed outside the framework and passed in as cell arrays of the decomposed sections as one would previously. 

Also corrects spelling of `canceled` to `cancelled` throughout.

### CODE REVIEW

**Review by cmarooney-stfc — APPROVED** [review]

File: herbert_core/classes/MPIFramework/@MessagesFilebased/private/check_job_cancelled_.m [hunk 1]
```diff
@@ -0,0 +1,13 @@
+function [is,err_code,err_message] = check_job_cancelled_(obj)
+% check if the filebased job is cancelled or message "cancelled" has been
+% send from the task with tid requested.
```
[the remark this answers is not in the document]
Comment (oerc0122) [reply herbert_core/classes/MPIFramework/@MessagesFilebased/private/check_job_cancelled_.m:3]: Agreed, I've sorted this in another function. Not quite sure where the tid (task ID?) is specified for the function.

File: herbert_core/classes/MPIFramework/messages/InitMessage.m [hunk 2]
```diff
@@ -53,7 +53,8 @@
             if ~exist('n_first_step', 'var')
                 n_first_step = 1;
             end
-            if iscell(loop_data)
+
```
[the remark this answers is not in the document]
Comment (oerc0122) [reply herbert_core/classes/MPIFramework/messages/InitMessage.m:56]: Yes, I considered rewriting this with an input parser, but I also wasn't sure what was default. I may still do this as it's not exactly clear what's happening.

File: herbert_core/classes/MPIFramework/messages/InitMessage.m [hunk 3]
```diff
@@ -37,22 +37,16 @@
...
-            obj.payload = struct('common_data',common_data,...
-                'loopData',[],'n_first_step',1,'n_steps',0,...
-                'return_results',return_results );
-            if ~exist('n_first_step', 'var')
-                n_first_step = 1;
-            end
+            p = inputParser();
+            addOptional('common_data', [])
+            addOptional('loop_data', 1)
+            addOptional('return_results', false, @isboolean)
```
Comment (hsautessella) [inline herbert_core/classes/MPIFramework/messages/InitMessage.m]: I think this should be `islogical`, or you could use Herbert's `islognumscalar`, which permits `0` and `1` (doubles) as well as `true` and `false`

File: herbert_core/classes/MPIFramework/messages/InitMessage.m (hunk 2 above)
[the remark this answers is not in the document]
Comment (nickbattam-tessella) [reply herbert_core/classes/MPIFramework/messages/InitMessage.m:51]: Would extracting this complex parser a function make it clearer?  That function can note clearly what defaults will be returned.

The only code left in the constructor then would be the assignment of data from the parsed args to the payload struct (which can be laid out vertically as `key, value, ...` and be clear how the args were then mapped to the payload

File: herbert_core/classes/MPIFramework/@JobExecutor/private/log_progress_.m [hunk 4]
```diff
@@ -10,70 +10,71 @@ function  log_progress_(obj,step,n_steps,time_per_step,add_info)
 %              job log
 % Outputs:
 % Sends message of type LogMessage to the job dispatcher.
-% Throws JOB_EXECUTOR:canceled error in case the job has been canceled.
-% Throws JOB_EXECUTOR:canceled error in case the reduce_messages returned
+% Throws JOB_EXECUTOR:cancelled error in case the job has been cancelled.
+% Throws JOB_EXECUTOR:cancelled error in case the reduce_messages returned
```
Comment (nickbattam-tessella) [inline herbert_core/classes/MPIFramework/@JobExecutor/private/log_progress_.m:14]: the linebreak before 'Failed message' makes this comment hard to parse

File: herbert_core/classes/MPIFramework/@JobDispatcher/private/split_tasks_.m [hunk 5]
```diff
@@ -7,105 +7,115 @@
...
+    for task_id=1:n_workers
+        if is_list
+            task_par_nums = (n_alloc_tasks(task_id)+1):(n_alloc_tasks(task_id)+each_task_npar(task_id));
+            worker_par_list{task_id}  = task_par_nums;
+        else
+            worker_par_list{task_id} = [(n_alloc_tasks(task_id)+1),each_task_npar(task_id)];
+        end
     end
-end
+end
```
Comment (nickbattam-tessella) [inline herbert_core/classes/MPIFramework/@JobDispatcher/private/split_tasks_.m:121]: newline

File: herbert_core/classes/MPIFramework/@JobDispatcher/private/split_tasks_.m [hunk 6]
```diff
@@ -7,105 +7,115 @@
...
         is_list = true;
-    else
-        error('JOB_DISPATCHER:invalid_argument',...
-            ' job loop parameters contains empty structure');
+
+    elseif isstruct(job_param_list) % array of structures
+
+        fn = fieldnames(job_param_list);
+        if ~isempty(fn)
+            n_tasks = numel(job_param_list.(fn{1}));
```
Comment (nickbattam-tessella) [inline herbert_core/classes/MPIFramework/@JobDispatcher/private/split_tasks_.m:77]: This is assuming something about  that the contents of the "first" field in the struct.
The documentation for `fieldnames` doesn't offer any guarantee what order fields will be returned.

File: herbert_core/classes/MPIFramework/@JobDispatcher/private/send_tasks_to_workers_.m [hunk 7]
```diff
@@ -36,93 +36,99 @@
...
-    fn = fieldnames(loop_param);
-    par1 = loop_param.(fn{1});
-    if iscell(par1)
-        n_jobs = numel(par1);
+    % Available number of workers
+    n_wk = n_workers;
+
+    if ~isscalar(loop_param)
+        n_jobs = numel(loop_param);
+    elseif isscalar(loop_param) && isnumeric(loop_param)
```
Comment (nickbattam-tessella) [inline herbert_core/classes/MPIFramework/@JobDispatcher/private/send_tasks_to_workers_.m:114]: is the `isscalar` test required here? The initial `if` will have caught any `loop_param` that are not scalar

**Review by abuts — APPROVED** [review]
looks ok to me. The main thing is taking on a chin the thankless task of renaming canceled to cancelled. 
A bit worry about using inputParser in InitMessage its absolutely non-obvious from looking at the code what would happen and if some parameters are missing. May be because low experience in using this function. Hope interface is guarded properly by unit tests.  As this is low priority improvement (you will test the job anyway), I do not think this remark should hold the deployment
